packages/polling-controller/src/PollingController.ts

```diff
@@ -0,0 +1,120 @@
+  Name extends string,
+  State extends Record<string, Json>,
+  messenger extends RestrictedControllerMessenger<
+    Name,
+    any,
+    PollingCompleteType<Name> | any,
+    string,
+    string
+  >,
+> extends BaseControllerV2<Name, State, messenger> {
+  readonly #intervalLength = 1000;
+
+  private readonly networkClientIdTokensMap: Map<NetworkClientId, Set<string>> =
+    new Map();
+
+  private readonly intervalIds: Record<NetworkClientId, NodeJS.Timeout> = {};
+
+  /**
+   * Starts polling for a networkClientId
+   *
+   * @param networkClientId - The networkClientId to start polling for
+   * @returns void
+   */
+  start(networkClientId: NetworkClientId) {
+    const innerPollToken = random();
+    if (this.networkClientIdTokensMap.has(networkClientId)) {
+      const set = this.networkClientIdTokensMap.get(networkClientId);
+      set?.add(innerPollToken);
+    } else {
+      const set = new Set<string>();
+      set.add(innerPollToken);
+      this.networkClientIdTokensMap.set(networkClientId, set);
+    }
+    this.#poll(networkClientId);
+    return innerPollToken;
+  }
+
+  /**
+   * Stops polling for all networkClientIds
+   */
+  stopAll() {
+    this.networkClientIdTokensMap.forEach((tokens, _networkClientId) => {
+      tokens.forEach((token) => {
+        this.stop(token);
+      });
+    });
+  }
+
+  /**
+   * Stops polling for a networkClientId
+   *
+   * @param pollingToken - The polling token to stop polling for
+   */
+  stop(pollingToken: string) {
+    if (!pollingToken) {
+      throw new Error('pollingToken required');
+    }
+    let found = false;
+    this.networkClientIdTokensMap.forEach((tokens, networkClientId) => {
+      if (tokens.has(pollingToken)) {
+        found = true;
+        this.networkClientIdTokensMap
+          .get(networkClientId)
+          ?.delete(pollingToken);
+        if (this.networkClientIdTokensMap.get(networkClientId)?.size === 0) {
+          clearTimeout(this.intervalIds[networkClientId]);
+          delete this.intervalIds[networkClientId];
+          this.networkClientIdTokensMap.delete(networkClientId);
+          this.messagingSystem.publish(
+            `${this.name}:pollingComplete`,
+            networkClientId,
+          );
+        }
+      }
+    });
+    if (!found) {
+      throw new Error('pollingToken not found');
+    }
+  }
+
+  /**
+   * Executes the poll for a networkClientId
+   *
+   * @param networkClientId - The networkClientId to execute the poll for
+   */
+  abstract executePoll(networkClientId: NetworkClientId): Promise<void>;
+
+  #poll(networkClientId: NetworkClientId) {
+    if (this.intervalIds[networkClientId]) {
+      clearTimeout(this.intervalIds[networkClientId]);
+      delete this.intervalIds[networkClientId];
+    }
+    this.intervalIds[networkClientId] = setTimeout(async () => {
+      try {
+        await this.executePoll(networkClientId);
+      } catch (error) {
+        console.error(error);
```

Comment:
Should we be eating the errors? I suppose even if we let the error bubble up, since this is happening inside an interval, nothing else would be able to catch the error anyways. In this case, does it make sense to add an interface to subscribe to errors? IE how would a controller using this abstract controller be able to detect an error in its executePoll call, and react to it.

Should we keep polling when there are errors? Maybe we can have an option to configure an linear/exponential backoff when there are errors.

Reply:
I had the same thought, this is definitely not ideal for error handling. But it's strictly better than the status quo, where typically the error is completely swallowed and not even logged. Something to consider later as an enhancement perhaps